File: apps/desktop/src/services/commands/updates.ts
import { invoke } from "@/services/tauri";

export type UpdateCheckMode = "auto" | "test";

export interface AppUpdateCheckResult {
  status: "not_configured" | "up_to_date" | "available";
  currentVersion: string;
  latestVersion: string | null;
  updateAvailable: boolean;
  source: string;
  message: string;
}

export interface ContentUpdatePackage {
  id: string;
  kind: "seed_db" | "data_patch" | "catalog" | "audio" | "asset";
  version: string;
  title: string;
  required: boolean;
  optional: boolean;
  sizeBytes: number;
  downloadPolicy: "auto" | "manual";
  audio: boolean;
}

export interface ContentUpdateCheckResult {
  status: "not_configured" | "available" | "empty" | "incompatible";
  manifestVersion: string | null;
  channel: string | null;
  source: string;
  totalPackages: number;
  requiredPackages: number;
  optionalAudioPackages: number;
  requiredDownloadBytes: number;
  optionalAudioBytes: number;
  packages: ContentUpdatePackage[];
  message: string;
}

export function checkAppUpdate(
  mode: UpdateCheckMode = "auto",
): Promise<AppUpdateCheckResult> {
  return invoke<AppUpdateCheckResult>("check_app_update", {
    input: { mode },
  });
}

export function checkContentUpdates(
  mode: UpdateCheckMode = "auto",
  manifestPath: string | null = null,
): Promise<ContentUpdateCheckResult> {
  return invoke<ContentUpdateCheckResult>("check_content_updates", {
    input: { mode, manifestPath },
  });
}

File: apps/desktop/src-tauri/src/dto/updates.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppUpdateCheckInputDto {
    pub mode: Option<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AppUpdateCheckResultDto {
    pub status: String,
    pub current_version: String,
    pub latest_version: Option<String>,
    pub update_available: bool,
    pub source: String,
    pub message: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContentUpdateCheckInputDto {
    pub mode: Option<String>,
    pub manifest_path: Option<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ContentUpdatePackageDto {
    pub id: String,
    pub kind: String,
    pub version: String,
    pub title: String,
    pub required: bool,
    pub optional: bool,
    pub size_bytes: u64,
    pub download_policy: String,
    pub audio: bool,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ContentUpdateCheckResultDto {
    pub status: String,
    pub manifest_version: Option<String>,
    pub channel: Option<String>,
    pub source: String,
    pub total_packages: usize,
    pub required_packages: usize,
    pub optional_audio_packages: usize,
    pub required_download_bytes: u64,
    pub optional_audio_bytes: u64,
    pub packages: Vec<ContentUpdatePackageDto>,
    pub message: String,
}
